### ISSUE TITLE
Feature: Add a search inside notes

### ISSUE THREAD

**Issue by rajeshisnepali** [standardnotes/forum#1770]
`OS: Ubuntu 20.04`

Need search feature inside a note that can be run by calling `Ctrl+F` .

There is an option to search text globally. But then I will have to find the card.

**Comment by henriqueffc** [comment]
I use the Standard Notes app, free version, on Arch Linux in AppImage format. The last update removed the option to search for text in notes. Please, was this change caused by some unexpected mechanism in the app or was it intentional and will it be permanent? The search option also no longer appears in the web version. I use Firefox. Thank you for developing this wonderful app.

Old version

![image](https://i.imgur.com/1XB2W4n.png)

New version

![image2](https://i.imgur.com/UR6iaSd.png)

**Comment by atmoio** [comment]
The search bar has been moved to the navigations/tags panel in the latest release.

This is to make way for searching more than just notes.

**Comment by henriqueffc** [comment]
Yup. Thank you for letting me know. I left that tab hidden.

![image](https://i.imgur.com/dUPV2w1.png)

**Comment by sethcasana** [comment]
The original issue was that it does not seem possible to find strings within a single note. Global search across multiple file types is great. However, not being able to restrict search to single text file is odd.

Additionally, there does not seem to be a way to advance through the hits of a global search. Instead, one must manually scroll through the notes and visually identify where the string was located.

Finally, when you use global search to discover a string, the results are given in the number of documents that contain at least one instance of that string instead of the number of times the string occurs. This is just confusing (see screen shot for example).

<img width="712" alt="Screen Shot 2022-07-14 at 5 48 57 AM" src="https://user-images.githubusercontent.com/173278/178954976-ccc45486-eb58-4099-815a-324ce45aa902.png">

These features have been rejected in the past with the explanation, "We don't want Standard Notes to become a text editor." Fair enough, but do you want it to be a note editor? Then allow users to limit text search to one document and easily jump to each hit within the document.

**Comment by sudokilleen** [comment]
> These features have been rejected in the past with the explanation, "We don't want Standard Notes to become a text editor."

What?! The entire purpose of a notes app is to edit text!
It's the sole failure of Standard Notes, and it's a massive one. Sure you can just use it in a browser to piggyback off its Ctrl+F, but it's not nearly as clean or convenient, especially on Android.
Without a proper search the actual apps are a complete waste of time to me. If it's not a "text editor", what is it? Productivity? This is the opposite of productive, and now I'm spending more time pleading for a basic **standard** of every ~~text editor~~ notes app made since the turn of the century. Is it Standard Notes or 'No Standards Notes'?

I guess I'll just have to set up some new browsers exclusively for a Standard Notes homepage with the URL bar removed. I'm fine with workarounds but a lot of people have dropped services for far less, and I can assure you this decision is losing them money because it's the only reason I won't go for Premium. I still love everything else about it/them, but there's really no excuse for this.

**Comment by avt00** [comment]
It is a must to have a text search inside a note.

**Comment by KevinSource0** [comment]
If the team is listening, please re-consider this feature. Search inside a note is an essential feature for note-taking app, given most people don't know / wouldn't take the time to report an issue, the number of people requested indicates this is high-demand, you're losing customers because of this.

**Comment by Saldef** [comment]
Yeah, This is a critical feature, especially for people who write long notes.
 
A workaround is to use the Code Editor with Vim mode enabled, then you can easily search inside of a note with highlights using the forward slash `/` in Vim's normal mode . But it is a bit inconvenient for people who don't prefer using Code Editor, as they need to switch to Code Editor each time they want to search inside of a note. 

I remember The Append Editor used to have it as well in its Monaco Editor. 

Since the devs are recently trying to minimize the number of Editors, I guess a solution for that would be to add the search functionality on each editor separately, in order not to compromise security.

**Comment by sethcasana** [comment]
It would be helpful if the developers could actually comment on the search within text feature. So far, there has been none addressing that specifically.

The per editor approach that @Saldef mentioned helps only when searching for strings within a single note. Using a two-step process of first a global search to find the relevant document and then a Editor-specific note search is a possible workaround, though the process seems clunky. It also requires users to know which editors offer internal string search and switching to them for each note.

If each editor mode were required to implement internal string search, that would go a long way to helping alleviate the problem without having to implement global internal string search as a top-level application feature.

**Comment by JimmysDev** [comment]
This has been requested for years and is badly needed. On iOS all you get is the note in which a word occurs, which is utterly useless much of the time in my experience. Highlighting seems like such an easy thing to implement, and such an odd choice to prohibit. You could still make it a setting that is off by default if you feel strongly about it, but this UX dictatorialism is not right, especially in the context of so many users requesting the feature.

**Comment by arekm** [comment]
Highlighting is not enough. If you have few pages long note (and I have such) you need to scroll over and over few times due to often missing what's highlighted (from practice; on a desktop version).

Marking on scroll bar would also be needed.

**Comment by adamhotep** [comment]
Please consider starting with the functionality in Firefox's find dialog (<kbd>Ctrl</kbd>+<kbd>f</kbd>). This supports:
* A pager that counts matches and lets you navigate up and down through them
* Toggle (default=on) to highlight all matches
* Toggle (default=off) to match case
* Toggle (default=off) to match diacritics (`resume` matches `résumé` unless this is on)
* Toggle (default=off) to limit search to whole words

Implementing diacritics is hopefully as easy as finding an existing library or implementing the [Unicode confusables](https://util.unicode.org/UnicodeJsps/confusables.jsp) (direct link: [confusables.txt](https://www.unicode.org/Public/security/8.0.0/confusables.txt). I'm not sure how Firefox does it. If that's hard, defer it. The other items are far more important.

I'd love to be able to search even more intelligently, like "foo NEAR bar" or with regular expressions, but let's cover the basics first. (This comes from #1700).

**Comment by ErinPo** [comment]
+1 very much looking forward to this.

**Comment by steel-steven-87** [comment]
What's the point of using a digital note taking app if you can't use the basic search that you can see in something as simple as Windows Notepad? Any time you have a larger note and you want to find something quickly, you have to skim through the whole page trying to find the text you want. In this case, why not switch to a paper notebook instead?

**Comment by torpax** [comment]
I also desperately need the ability to search within a note.

**Comment by SynapseScribe** [comment]
this should have greater priority..

**Comment by Techn1x** [comment]
~I had all my notes saved as notepad/text files on my desktop. I moved all my notes over to Standard Notes a few weeks ago, and now I can't search properly and my productivity has dropped~ :upside_down_face: 

Search/highlighting within notes is a basic _expected_ feature, it's even in Notepad, can't get more basic than that. It is not an advanced text editor feature...

~Glad my notes are secure now & synced now, but had I known beforehand that this feature was missing & not a priority, I would have reconsidered. I am currently a freeloader and was looking to upgrade for the markdown support - but I won't bother until this is given some form of priority to implement. Heck, make it a paid feature if it gets it across the line~ :smile: 

EDIT: I'm taking back my comments - because if you use the main notes search it does seem to _sometimes_ find the word & highlight. What's annoying is that CTRL+F doesn't focus the search field, and I end up typing my query into the note without realising. And when using the main field, it only sometimes navigates to the found entry - most times it remains at the top of the note. I'm using the ubuntu standard notes app

**Comment by eylenburg** [comment]
This is desperately needed and should be easy to implement

**Comment by sudokilleen** [comment]
Here is the easy workaround I've been using for months. It should work on any desktop OS (Win/Linux/ChromeOS):

- Open any [Chromium](https://download-chromium.appspot.com/)-based browser
- Go to app.standardnotes.com -> Click Chromium's 3-dot menu in the top right -> "More tools..." -> ["Create shortcut..."](https://user-images.githubusercontent.com/109898451/211897200-960c723b-5fac-4b10-8a29-7c2dea105c67.png) -> **Check** [**"Open as window"**](https://user-images.githubusercontent.com/109898451/211897454-e7964afb-d779-48c8-a940-fac8b5838b52.png) and then click "Create"

The web app is virtually identical to the real app **BUT WITH A CTRL+F!** It should automatically hide the browser's UI, use the Standard Notes icon in your taskbar, receive all the SN keyboard shortcuts, and everything else you get from the normal app.

![](https://user-images.githubusercontent.com/109898451/211893426-6c6d4135-9d12-4536-be72-e664d5b339eb.png)

_(You can also make a nice ProtonMail app this way, and you can use extensions, i.e. Spotify + uBlock for an ad-free Spotify app. If you're on Linux you can use the [Web App Manager](https://github.com/linuxmint/webapp-manager) with any browser (non-Chromium)_.

**Comment by SynapseScribe** [comment]
tbh i don't see any point in having the app installed on desktop, but that's up to each one's preference. I use app.standardnotes.com in a pinned tab in browser since forever and search notes with ctrl+F. this is naturally straightforward. the real problem is on mobile app (eg. on android)

**Comment by jjulius50** [comment]
SudoKilleen, thanks a lot! You taught me something I did not know and will definitely use with SN and PM.

**Comment by myreli** [comment]
edit: I've previously marked it as duplicated, but users pointed out that this is a different suggestion: to feature in-note search inside plaintext notes (and in the free tier). 

---

~~Duplicates https://github.com/standardnotes/forum/discussions/2498 (TL;DR: it's now available through Super note type)~~

**Comment by eylenburg** [comment]
The "Super" editor is only available for the premium plans. Really? You're locking "Ctrl + F" behind a subscription?

**Comment by sethcasana** [comment]
This has been marked as a duplicate, but it is not since the original problem still persists: users are unable to find search results within a plain text note. Indeed, many of the comments in this thread point out how other editor types allow for search and recommend it as a workaround.

In that linked feature request, there is a post from the Maintainer where he argues against adding a search feature based on the "longevity" principle of development:

https://standardnotes.com/longevity

How that stance squares with launching a subscription-only "Super" note escapes me. Regardless, what the developers have signaled here is that they do not think that find and replace is too complex to survive the apocalypse, but that you cannot have that feature for free. I trust that their marketing materials will be updated soon to reflect that value proposition. I have been a Standard Notes Extended subscriber for years, but at this point, I wish I hadn't.

**Comment by myreli** [comment]
Hi @eylenburg, @sethcasana and @fusionneur, 

Sorry, I might have misunderstood the suggestion as a duplicate, we can re-open it and leave it as a **suggestion for plaintext notes to also feature in-note search**. 

Please note I'm not a developer of the SN team, and I'm only trying to improve our forum experience by triaging our older issues so that the team can respond to them more effectively. I don't respond in name of the team but as part of the community. Since many subscribers were satisfied by the Super editor in other questions (mostly on Discord) I assumed the same could apply here. I apologize if anything came out wrong. 

From my understanding, the plain text note is meant to be future-proof, and that is why its implementation is as simple as possible: a standard HTML textarea input. This means that in-note search works through the browser implementation, as [shared here](https://github.com/standardnotes/forum/issues/1770#issuecomment-1379391803). There is also the free alternative through the [community plugins](https://github.com/jonhadfield/awesome-standard-notes), I for instance used the Fast Editor before super. 

In the issue I linked, which is much older, SN had a different perspective on how the product was shaped. Vision has quite changed since then on what is built-in and what is extended, that's why Files, Folders and other features became native. You can read about this decision in [SN's blog](https://standardnotes.com/blog/folders-from-extension-to-native). 

Hopefully, I could clarify my intention with triaging issues and sorry for any disruption. I will take note of how I respond to/close other requests. Thanks for the input.

**Comment by sethcasana** [comment]
Thank you so much for your response, @myreli. Your comments reflect the same sentiments that keep coming up in this thread, that being, search seems to work in every implementation except default plain text notes. The workarounds you mention are just that—workarounds.

The newer blog post you linked to mentioned that SN wants to move away from extensions so that the default experience of the app is as pleasant and useful as possible. Implementing search functionality in plain text notes would go a long way towards that goal. I and the others who have commented here would  greatly appreciate any assistance you can offer in that regard.

**Comment by atmoio** [comment]
Plain text cannot offer search in note because it is an HTML textarea and offers 0 points of customization. If it were possible we would have built it long ago. Super came as a premium editor many months before us realizing we could even build search for it.

**Comment by SynapseScribe** [comment]
is there any update on adding search within note on android app?

**Comment by effieeee** [comment]
Hi @fusionneur, as Mo mentioned above, I'm afraid we cannot add search within notes for Plain Text, though the search feature for **Super** notes is already accessible on mobile! On the formatting toolbar above the keyboard, scroll a bit to the right to find the **magnifying glass** icon, which will open the search field:

![image](https://github.com/standardnotes/forum/assets/54989748/3c526388-ea51-4e3d-8f70-47826250f8ea)

**Comment by lesderid** [comment]
> Plain text cannot offer search in note because it is an HTML textarea and offers 0 points of customization. If it were possible we would have built it long ago. Super came as a premium editor many months before us realizing we could even build search for it.

Honestly this is simply not true. There's plenty of ways to add search to a `textarea`. I get that the SN team doesn't want to do it, although I disagree with that decision, but I feel like it's a bit disingenious to claim it's because of technical reasons.

Can we please just get a decisive statement from the team and either close the issue or edit it to track a feature that the team *is* willing to implement?

**Comment by JimmysDev** [comment]
Even if another layer is required to enable search, it is such a vital feature that it would be worth the added weight.

**Comment by danielmanesku** [comment]
I have a subscription that would allow me to use Supernotes. However, most of my notes are already in Markdown. Even converting to Supernotes is not trivial as I will lose a bunch of formatting. I don't want to spend hours to adjust formatting on ~100 of my notes.

Please add a simple search to regular notes.

**Comment by BarRunBoatHonestyOperation** [comment]
Is there a workaround for iOS for searching/highlighting within notes, maybe with an extension? I'm often out away from my computer and need to look up a specific text field but search only gives me the note title and that's it. I've gotten very good at scrolling and speed-reading...

Or is migrating all my notes to Super Notes the only option? That's the only feature I need from Super Notes so converting all my notes to Super Notes seems like a hassle.

**Comment by SynapseScribe** [comment]
is there any update? it's quite ridiculous to have a notes app without being able to search inside the note. even if you open the good old windows notepad you can search it..

**Comment by Yolakalemowa** [comment]
> is there any update? it's quite ridiculous to have a notes app without being able to search inside the note. even if you open the good old windows notepad you can search it..

still waiting :'( 

this issue tagged "triage" and "enhancement" . I'm not familiar with github much, does that mean anything to you?

**Comment by raacampbell** [comment]
I would love ctrl+F search in plain text notes. It's a critical feature in notes with a lot of content. I have the paid subscription but what about notes that I want to keep in MarkDown?

**Comment by hoerdter** [comment]
Not being able to search my notes on android is a huge bummer... The app is fantastic otherwise.

**Comment by virtualo1** [comment]
Wow, this is unreal. This is all the Standard Notes is missing. Searchability!

Simple Notes implements the search function perfectly. Look what they did for inspiration on coding this. I have since switched away from Simple Notes due to lack of security.

I am now with Joplin however it is a bit clunky. I really wanted to switch off Joplin to Standard Notes but no searchability is a deal breaker.

**Comment by Yolakalemowa** [comment]
> Wow, this is unreal. This is all the Standard Notes is missing. Searchability!
> 
> Simple Notes implements the search function perfectly. Look what they did for inspiration on coding this. I have since switched away from Simple Notes due to lack of security.
> 
> I am now with Joplin however it is a bit clunky. I really wanted to switch off Joplin to Standard Notes but no searchability is a deal breaker.

Agreed. It's ridiculous. The only reason I haven't switched from SN yet (I'm thinking about Obsidian) is due to pure laziness. I'm tired of switching from one note app to the next.

**Comment by fabiosalles** [comment]
Humanity is on the verge of placing people on Mars - the best forecast is FOUR years from now. We shall be stomping that red dust before 2030.

If we can send a man up to Mars, ANYTHING gotta be possible. It is even a common trope nowadays - you can be anything, you can do anything etc. etc. etc.

But we can't have FIND and FIND NEXT buttons inside a gizmo that now is an almost literal _The Hitchhikers Guide's To The Galaxy_?! Can't be more difficult than finding cancer's cure!

No, for freaking out loud, no.

Get it done, please. Show some deep craftsmanship, some cunning thinking and creative fiddling and just do it.

**Comment by marcosdiasalves** [comment]
Just announced: [Adds in-note search for plain text notes](https://github.com/standardnotes/app/releases/tag/%40standardnotes%2Fdesktop%403.202.0)

**Comment by Yolakalemowa** [comment]
> Just announced: [Adds in-note search for plain text notes](https://github.com/standardnotes/app/releases/tag/%40standardnotes%2Fdesktop%403.202.0)

finally!!! @fabiosalles ... next grand achievement hopefully before mars landing should be notesnook figuring out how to switch from LTR to RTL seemlessly lmao.

**Comment by SynapseScribe** [comment]
> Just announced: [Adds in-note search for plain text notes](https://github.com/standardnotes/app/releases/tag/%40standardnotes%2Fdesktop%403.202.0)

will this be allowed on free plan as well? thank you

**Comment by Yolakalemowa** [comment]
> > Just announced: [Adds in-note search for plain text notes](https://github.com/standardnotes/app/releases/tag/%40standardnotes%2Fdesktop%403.202.0)
> 
> will this be allowed on free plan as well? thank you

yes it is. but looks like you can only search INSIDE each note. 
But you can't like search for a keyword and be shown all results containing that word in ALL notes. (which would be ideal) 
but hey better than nothing...

**Comment by SynapseScribe** [comment]
on android app i go to notes on  the side panel, then search and shows me the notes with that keyword. it won't show the line or a snippet of text where it is, but together with the new search feature within a note, that's enough for me at least

**Comment by fabiosalles** [comment]
I have just tested it and it works nicely!! Thanks, Standard Notes Team! :-)
